### PR TITLE
ARCHI-338 add missing environment reference in ApiKey, Subscription and plan

### PR DESCRIPTION
**Pull request by guillaumelamirand**
## Issue

https://gravitee.atlassian.net/browse/ARCHI-338

## Description

This PR adds a extra attribute in ApiKey, Subscription and Plan repository in order to be able to filter them according to the right tenant.

This is required to secure any call made through the bridge and will solve a major issue when all ApiKey and Subscription are sent to any gateways. Meaning that in Hybrid mode a gateway will see all ApiKey of all Environnement.

The repository are upgrading using native upgrader (liquibase or mongodb).

## Additional context

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-btkcuhbktd.chromatic.com)

### CODE REVIEW

File: gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/plans/ApiIndexUpgrader.java [hunk 1]
```diff
@@ -32,6 +32,7 @@ protected Index buildIndex() {
             .collection("plans")
             .name("a1")
             .key("api", ascending())
+            .key("environmentId", ascending())
```
Comment (phiz71) [inline gravitee-apim-repository/gravitee-apim-repository-mongodb/src/main/java/io/gravitee/repository/mongodb/management/upgrade/upgrader/index/plans/ApiIndexUpgrader.java]: As each upgrader is run only once, I don't know how to force this upgrader to be executed to update an existing index.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 import io.gravitee.apim.core.audit.model.AuditActor;
 import io.gravitee.apim.core.audit.model.AuditInfo;
+import io.gravitee.apim.core.plan.model.Plan;
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-management-v2/gravitee-apim-rest-api-management-v2-rest/src/main/java/io/gravitee/rest/api/management/v2/rest/resource/api/ApiPlansResource.java]: Is it needed ?

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/exceptions/SubscriptionMismatchEnvironmentException.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    private final String planId;
+
+    public SubscriptionMismatchEnvironmentException(final String applicationId, final String planId) {
+        this.applicationId = applicationId;
+        this.planId = planId;
+    }
+
+    @Override
+    public String getMessage() {
+        return ("Application [" + applicationId + "] and plan [" + planId + "] doesn't bellong to the same environment");
```
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/exceptions/SubscriptionMismatchEnvironmentException.java]:
```suggestion
        return ("Application [" + applicationId + "] and plan [" + planId + "] doesn't belong to the same environment");
```

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PlanConverter.java
Comment (phiz71) [inline gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PlanConverter.java:1]: Why don't you set the environmentId when converting from a PlanEntity to a Plan?
```java
public io.gravitee.definition.model.Plan toPlanDefinition(PlanEntity plan) {
```

**Review by phiz71 — CHANGES_REQUESTED** [review]
I'm globally ok with your PR.
But I have some concerns about its impact. I can see 3 topics in particular:

1. You are modifying the interface of the repository-api for `PlanRepository`. The name and the signature of the method change.
Will it be a problem for the bridge, assuming that bridge 4.4 must be able to response to bridge clients in version 4.1, 4.2 & 4.3?

2. Feature parity team has worked on the import feature for the 4.4.0. I suspect some conflicts around the plan management. If it's not already the case, I would suggest to have a meeting with them so you can sync your respective work.

3. I saw that you have added the environmentId also in the "xxxEntity" objects. But I'm not sure to see if it's really needed or not. I would have thought that the environmentId was defined at creation time only, and then was never required in the service layer. But I may be wrong. Maybe we can have a quick chat for that point.

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PlanConverter.java
Comment (guillaumelamirand) [reply gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/converter/PlanConverter.java:1]: The io.gravitee.definition.model.Plan is for the gateway, no need of the environment.

**Review by phiz71 — APPROVED** [review]

**Pull request merged**